kratos/processes/calculate_distance_to_skin_process.cpp

```diff
@@ -7,215 +7,228 @@
 //  License:		 BSD License
 //					 Kratos default license: kratos/license.txt
 //
-//  Main authors:    Pooyan Dadvand
+//  Main authors:    Pooyan Dadvand, Ruben Zorrilla
 //
 
-
 // System includes
 
-
 // External includes
 
-
 // Project includes
 #include "geometries/plane_3d.h"
 #include "processes/calculate_distance_to_skin_process.h"
 #include "utilities/geometry_utilities.h"
-
+#include "utilities/intersection_utilities.h"
 
 namespace Kratos
 {
 
-	CalculateDistanceToSkinProcess::CalculateDistanceToSkinProcess(ModelPart& rVolumePart, ModelPart& rSkinPart)
-		: CalculateDiscontinuousDistanceToSkinProcess(rVolumePart, rSkinPart)
+	template<std::size_t TDim>
+	CalculateDistanceToSkinProcess<TDim>::CalculateDistanceToSkinProcess(
+		ModelPart& rVolumePart, 
+		ModelPart& rSkinPart)
+		: CalculateDiscontinuousDistanceToSkinProcess<TDim>(rVolumePart, rSkinPart)
 	{
 	}
 
-	CalculateDistanceToSkinProcess::~CalculateDistanceToSkinProcess()
+	template<std::size_t TDim>
+	CalculateDistanceToSkinProcess<TDim>::~CalculateDistanceToSkinProcess()
 	{
 	}
 
-	void CalculateDistanceToSkinProcess::Initialize()
+	template<std::size_t TDim>
+	void CalculateDistanceToSkinProcess<TDim>::Initialize()
 	{
-		CalculateDiscontinuousDistanceToSkinProcess::Initialize();
+		CalculateDiscontinuousDistanceToSkinProcess<TDim>::Initialize();
 		this->InitializeNodalDistances();
 	}
 
-	void CalculateDistanceToSkinProcess::InitializeNodalDistances()
+	template<std::size_t TDim>
+	void CalculateDistanceToSkinProcess<TDim>::InitializeNodalDistances()
 	{
-		ModelPart& ModelPart1 = (CalculateDiscontinuousDistanceToSkinProcess::mFindIntersectedObjectsProcess).GetModelPart1();
+		// Get the volume model part from the base discontinuous distance process
+		ModelPart& ModelPart1 = (CalculateDiscontinuousDistanceToSkinProcess<TDim>::mFindIntersectedObjectsProcess).GetModelPart1();
 
-		for (auto& node : ModelPart1.Nodes())
-		{
+		// Initialize the nodal distance values to a maximum positive value
+		for (auto& node : ModelPart1.Nodes()){
 			node.GetSolutionStepValue(DISTANCE) = std::numeric_limits<double>::max();
 		}
 	}
 
-	void CalculateDistanceToSkinProcess::CalculateDistances(std::vector<PointerVector<GeometricalObject>>& rIntersectedObjects)
+	template<std::size_t TDim>
+	void CalculateDistanceToSkinProcess<TDim>::CalculateDistances(
+		std::vector<PointerVector<GeometricalObject>>& rIntersectedObjects)
 	{
-		this->CalculateElementalDistances(rIntersectedObjects);
+		// Compute the discontinuous (elemental) distance field
+		const bool use_base_elemental_distance = false;
```

Comment:
@pooyan-dadvand I though that it is a good idea to let the use of both the base and the simple (implemented in this class) elemental distance computation. The use of the naive one (without plane optimization) is hardcoded. This is the one that we've used historically and it yields smoother continuous distance fields according to the preliminary tests I've done.